tracing/src/test/java/com/palantir/tracing/TracersTest.java

```diff
@@ -450,6 +457,292 @@ public void testWrapRunnableWithAlternateTraceId_traceStateRestoredToCleared() {
         assertThat(Tracer.hasTraceId()).isFalse();
     }
 
+    @Test
+    public void testWrappingFutureCallback_futureCallbackTraceIsIsolated() throws Exception {
+        FutureCallback<Void> futureCallback = new FutureCallback<Void>() {
+            @Override
+            public void onSuccess(@NullableDecl Void result) {
+                Tracer.startSpan("inside"); // never completed
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                Tracer.startSpan("inside"); // never completed
+            }
+        };
+
+        ListeningExecutorService listeningExecutorService =
+                MoreExecutors.listeningDecorator(Executors.newSingleThreadExecutor());
+        ListenableFuture<Void> success = listeningExecutorService.submit(() -> null);
+        ListenableFuture<Void> failure = listeningExecutorService.submit(() -> {
+            throw new IllegalStateException();
+        });
+
+        Tracer.startSpan("outside");
+        FutureCallback<Void> successCallback = Tracers.wrap(futureCallback);
+        Futures.addCallback(success, successCallback, MoreExecutors.directExecutor());
+        success.get();
+        assertThat(Tracer.completeSpan().get().getOperation()).isEqualTo("outside");
+
+        Tracer.getAndClearTrace();
+        Tracer.startSpan("outside");
+        FutureCallback<Void> failureCallback = Tracers.wrap(futureCallback);
+        Futures.addCallback(failure, failureCallback, MoreExecutors.directExecutor());
+        assertThatThrownBy(failure::get).isInstanceOf(ExecutionException.class);
+        assertThat(Tracer.completeSpan().get().getOperation()).isEqualTo("outside");
+    }
+
+    @Test
+    public void testWrappingFutureCallback_traceStateIsCapturedAtConstructionTime() throws Exception {
+        FutureCallback<Void> futureCallback = new FutureCallback<Void>() {
+            @Override
+            public void onSuccess(@NullableDecl Void result) {
+                assertThat(Tracer.completeSpan().get().getOperation()).isEqualTo("before-construction");
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                assertThat(Tracer.completeSpan().get().getOperation()).isEqualTo("before-construction");
+            }
+        };
+
+        ListeningExecutorService listeningExecutorService =
+                MoreExecutors.listeningDecorator(Executors.newSingleThreadExecutor());
+        ListenableFuture<Void> success = listeningExecutorService.submit(() -> null);
+        ListenableFuture<Void> failure = listeningExecutorService.submit(() -> {
+            throw new IllegalStateException();
+        });
+
+        Tracer.startSpan("before-construction");
+        FutureCallback<Void> successCallback = Tracers.wrap(futureCallback);
+        Tracer.startSpan("after-construction");
+        Futures.addCallback(success, successCallback, MoreExecutors.directExecutor());
```

Comment:
We should have a test that runs on an executor using a different thread, since tracing-java heavily uses thread state. The direct executor just invokes `run()` on the current thread.